New module with Github actions

jsirianni opened the pull request:
### What does this PR accomplish?

This PR is at the request of @qingling128. This module brings Puppet support for the Google Ops Agent and more (see README.md for details). 

### Continuous Integration

Github actions will test all agents on Linux and the Ops Agent on Windows.

![Screenshot from 2021-06-22 11-29-05](https://user-images.githubusercontent.com/23043836/122954019-1567d180-d34d-11eb-855d-07d8cf8cc267.png)

```bash
  Service google-cloud-ops-agent.target
     ✔  is expected to be installed
     ✔  is expected to be enabled
     ✔  is expected to be running
  System Package google-cloud-ops-agent
     ✔  is expected to be installed
     ✔  version is expected to match /1.0.4/
  File /etc/google-cloud-ops-agent/config.yaml
     ✔  is expected to exist
```

A handful of prerequisites are required before CI will finish successfully. I can assist in this area. See `test/README.md` for specific details.

Commit:
Replace (#1)

* replace repo with new puppet module

* update workflows to reflect new branch names

Commit:
fix badge links

Commit:
Add windows ci (#2)

* Add windows ci

* fix branch

* add windows image

* rename because we now support windows

* test windows

Commit:
Windows CI: 1.0.4 --> 1.0.10 (#3)

* switch from 1.0.4 to 1.0.10 on windows, the agent stop bug is resolved sometime after 1.0.4

* test 1.0.10 in github actions

* now that stopping a service works, re-add custom_config testing

* fix path to source config

* Fix resource name typo

* try running as system user. this command should work but fails in ci. it works under rdp when running 'as administrator'

* revert back to command prompt now that the new agent can stop currectly. powershell syntax keeps failing for some reason

* revert user: SYSTEM. puppet on windows cannot switch users, but I have comfirmed that the ci user should be able to stop services without being elevated

* sleep before restarting the agent to avoid a race condition where we restart during the intiail agent startup (results in a windows error)

* puppet does not like the >NUL param

* fix inspec test now that puppet windows restarts are working

* checksum seems to change on the windows filesystem, test the checksum for install and custom_install, using the hash returned from inspec failures

* correct checksum for install

qingling128 commented on .github/workflows/linux.yml:
```diff
@@ -0,0 +1,150 @@
+          - logging
+          - monitoring
+        version:
+          #- "1.0.4"
```
Is this line commented out by accident? Looks like the `exclude` section below is for `1.0.4` and the `test/cases/linux/ops-agent/1.0.4` folder has a bunch of files. So I guess it's still needed.

BTW, do we have any plan to add tests (Does not have to be this PR) for `MAJOR.*.*` versions?

jsirianni replied:
No, it should not have been commented, I have enabled it.

We can add additional version tests, such as 1.*.*, however, the matrix will become very large. I am okay with this, just needs to be considered.

qingling128 replied:
`MAJOR.*.*` is actually the recommended way out of the 3 options. So let's add tests for it (Follow up PR is fine). If the concern is about cost of testing, we could follow up offline to ensure it is handled by us.

qingling128 replied:
Filed https://github.com/GoogleCloudPlatform/puppet-google-logging/issues/9 for tracking. Resolving this thread.

qingling128 commented on .github/scripts/inspec.sh:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
```
[Can be done in a follow up PR as well]

Can we add a file header / license checker (e.g. https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/blob/master/.github/header-checker-lint.yml) and headers to each applicable files?

Sample header:

```
#!/bin/bash

# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Package build scripts can use this by setting $DESTDIR before launching the
# script.
```

jsirianni replied:
I have added a CI workflow that will fail a build when a license header is missing. I also added license headers to all source files / scripts.

qingling128 replied:
Great. Thanks!

qingling128 commented on .github/scripts/inspec.sh:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+
```
[Can be done in a follow up PR as well]

Also could we add a brief comment / description to each file to summarize what it does / how it fits into the bigger picture? This helps developers in the future to ramp up with the codebase more smoothly.

jsirianni replied:
Done 👍

qingling128 commented on manifests/agent.pp:
```diff
@@ -1,31 +1,170 @@
+
+  if $facts['os']['family'] == 'windows' {
+    if $installed {
+      exec { "install-repo-${agent_type}":
```
[Not necessarily need to change anything]

Just a side note: We now have an installation script for Windows as well: https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/installation#install-specific-version.

jsirianni replied:
I saw this. I think this is a future item. The package manager is working well right now. The script would be a bit more portable.

qingling128 replied:
Sounds good.

qingling128 commented on metadata.json:
```diff
@@ -1,87 +1,67 @@
-        "2016 R2",
-        "2016 R2 Core"
+        "2019",
+        "10"
```
The Windows operating systems we support are (https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#windows_operating_systems):

```
Windows Server: windows-2012-r2, windows-2016, windows-2019
Windows Server Core: windows-2012-r2-core, windows-2016-core, windows-2019-core
```

It would be great for the tests to match that list.

jsirianni replied:
This is certainly possible. Configuring the test images for windows will take some time. I would like to add this in a future PR if possible.

qingling128 replied:
Yup, that sounds good. We could explore a bit more for whether there is a way to avoid maintaining custom images.

Filed https://github.com/GoogleCloudPlatform/puppet-google-logging/issues/10 for tracking. Resolving this.

qingling128 commented on Rakefile:
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
```
[Probably a follow up PR]

Can we add some instructions around:

* Development workflow (e.g. if there is any local tests that developers could run instead of doing through the full e2e.)
* Release instructions (e.g. The changelog util here seems to expect specific best practices)

?

jsirianni replied:
I agree and think this would be best for a follow up PR.

qingling128 replied:
Resolving this. Filed https://github.com/GoogleCloudPlatform/puppet-google-logging/issues/11 for tracking.

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
-# Google Stackdriver Logging Puppet Module
+# Cloud Operations Puppet Agents
```
```suggestion
# Google Cloud Operations Agents Puppet Integration
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
-To install this module on your Puppet Master (or Puppet Client/Agent), use the
-Puppet module installer:
+- Linux
+  - [Cloud Ops Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent)
```
```suggestion
  - [Ops Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent)
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
-Puppet module installer:
+- Linux
+  - [Cloud Ops Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#supported_vms)
```
```suggestion
    - [Supported operating systems](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#supported_operating_systems)
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
+  - [Cloud Ops Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#supported_vms)
+  - [Monitoring Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/monitoring)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/monitoring#supported_vms)
```
```suggestion
    - [Supported Operating Systems](https://cloud.google.com/stackdriver/docs/solutions/agents/monitoring#supported_operating_systems)
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
+  - [Monitoring Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/monitoring)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/monitoring#supported_vms)
+  - [Logging Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/logging)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/logging#supported_vms)
```
```suggestion
    - [Supported Operating Systems](https://cloud.google.com/stackdriver/docs/solutions/agents/logging#supported_operating_systems)
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/logging#supported_vms)
+- Windows
+  - [Cloud Ops Agent](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent)
+    - [Supported Instances](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#supported_vms)
```
```suggestion
    - [Supported Operating Systems](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#supported_operating_systems)
```

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
 
-Optionally you can install support to _all_ Google Cloud Platform products at
-once by installing our "bundle" [`google-cloud`][bundle-forge] module:
+Permissions to the Google Cloud API. If you are running an old Compute Engine instance or Compute Engine instances created without the default credentials, then you must complete the following steps https://cloud.google.com/monitoring/agent/authorization#before_you_begin.
```
```suggestion
https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent#access
```

I thought most of the old Compute Engine instances mentioned in `If you are running an old Compute Engine instance or Compute Engine instances created without the default credentials, then you must complete the following steps https://cloud.google.com/monitoring/agent/authorization#before_you_begin.` should be deprecated by now. Did we run into similar issues and needed to do this?

jsirianni replied:
Sounds good. I did not run into any issues, but I did pull this from the Ansible repo.

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
+| Parameter               | Default       | Description                                                       |
+| ---                     | ---           | ---                                                               |
+| `agent_type`            | Required      | The agent type: `ops-agent`, `monitoring`, `logging`              |
+| `installed`             | `true`        | Boolean value. Ensure that the agent is installed or uninstalled  |
```
Could we do something similar to the Ansible implementation here? aka `package_state` (`present` and `absent`) instead of `installed` (true / false)?

For the context, we had a long discussion about this when working on Ansible. One of the benefits are to be forward compatible if in the future we need to support more package state (e.g. always upgraded to the latest) and service state (e.g. running / stopped).

jsirianni replied:
I agree with the consistency here, done 👍

qingling128 commented on README.md:
````diff
@@ -1,68 +1,155 @@
+
+Install the latest version:
+```ruby
+cloud_ops::agent {'ops-agent':
````
We have some branding guidelines internally to use `google_cloud_ops` for certain cases. Could we adjust this module name?

```suggestion
google_cloud_ops::agent {'ops-agent':
```

jsirianni replied:
Resolved https://github.com/GoogleCloudPlatform/puppet-google-logging/pull/8/commits/3b157511d7da4fc8cfd74f486a7aa5da1bd17e3d

qingling128 commented on README.md:
```diff
@@ -1,68 +1,155 @@
+| `agent_type`            | Required      | The agent type: `ops-agent`, `monitoring`, `logging`              |
+| `installed`             | `true`        | Boolean value. Ensure that the agent is installed or uninstalled  |
+| `version`               | `latest`      | The version variable can be used to specify which version of the agent to install. The allowed values are latest, MAJOR_VERSION.*.* and MAJOR_VERSION.MINOR_VERSION.PATCH_VERSION, which are described in detail below. |
+| `main_config`           |               | Optional value for overriding the default configuration           |
```
```suggestion
| `main_config`           |               | Optional value for overriding the default configuration. For configuration syntax instructions, see [Ops Agent Config](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/configuration), [Logging Agent Config](https://cloud.google.com/logging/docs/agent/logging/configuration), and [Monitoring Agent](https://cloud.google.com/monitoring/agent/monitoring/configuration for configuration) for more details.           |
```

qingling128 commented on metadata.json:
```diff
@@ -1,87 +1,67 @@
-  "version": "0.1.3",
-  "author": "Google",
-  "summary": "A Puppet module to manage Google Stackdriver Logging Agent",
+  "name": "ops-cloud_ops",
```
```suggestion
  "name": "google-cloud-ops-agents-puppet",
```

jsirianni replied:
I went with `google-cloud_ops_agents`, I believe this field is related to Puppet Forge <username>-<module name>. Module name can contain underscores. 

Or should we continue with the convention used in the existing metadata? https://github.com/GoogleCloudPlatform/puppet-google-logging/blob/master/metadata.json

qingling128 commented on metadata.json:
```diff
@@ -1,87 +1,67 @@
+  "name": "ops-cloud_ops",
+  "version": "0.1.0",
+  "author": "Google Inc.",
+  "summary": "Manage Google Cloud Ops Agent",
```
```suggestion
  "summary": "Manage Google Cloud Ops Agents",
```

qingling128 commented on metadata.json:
```diff
@@ -1,87 +1,67 @@
-  "source": "https://github.com/GoogleCloudPlatform/puppet-google-logging",
-  "project_page": "https://github.com/GoogleCloudPlatform/puppet-google-logging",
-  "issues_url": "https://github.com/GoogleCloudPlatform/puppet-google-logging/issues",
+  "source": "https://github.com/BlueMedora/google-puppet-agents",
```
`https://github.com/GoogleCloudPlatform/google-cloud-ops-agents-puppet` (We plan to rename this repo to `google-cloud-ops-agents-puppet`)?

qingling128 commented on test/README.md:
```diff
@@ -0,0 +1,166 @@
+
+Packer will require access to Google APIs. This [guide provides](https://cloud.google.com/build/docs/building/build-vm-images-with-packer) examples on how to setup your working environment.
+
+### Linux
```
Are these images built once and hosted somewhere, or built from scratch every time during a test run? I guess for Windows it is probably hosted somewhere, since it's built manually, right?

jsirianni replied:
I pushed a commit clarifying this. Basically, Packer is used to prepare base images for Linux. The same process is done manually for Windows. Terraform will deploy these "baked" images for Linux and Windows tests every time CI runs. The Windows instances are not left behind to be re-used, they are created on demand.

The project being used for CI should own these images, or have read access to them if they are in a different project.

qingling128 commented on test/cases/linux/ops-agent/1.0.4/custom_config/config.yaml:
```diff
@@ -0,0 +1,45 @@
+# Copyright 2020 Google LLC
```
Are the test files for `1.0.4` maintained by some automation system (aka auto generated), or entered by hand?

jsirianni replied:
They are not generated, created by hand. This is okay for now but more complex CI should probably have some sort of generation process.

qingling128 commented on test/cases/linux/ops-agent/1.0.4/uninstall/manifests/site.pp:
```diff
@@ -0,0 +1,5 @@
+cloud_ops::agent {'ops-agent':
+  agent_type => 'ops-agent',
+  version    =>  '1.0.5',
```
Does this intended to be `1.0.4`? Similar below, there are a few occurrences of `1.0.5`.

jsirianni replied:
This is resolved by replacing 1.0.4 with 2.0.1.

qingling128 commented on test/cases/linux/logging/latest/custom_config/files/plugins/plugin.rb:
```diff
@@ -0,0 +1,17 @@
+module Fluent
```
If we are looking for a sample fluentd config for 3rd party app (In some cases we call it `plugin` but it's actually not very accurate), here is an example: https://github.com/GoogleCloudPlatform/fluentd-catch-all-config/blob/master/configs/config.d/apache.conf

jsirianni replied:
Understood, I have implemented the config used here https://github.com/GoogleCloudPlatform/google-cloud-ops-agents-ansible/blob/master/molecule/resources/example_configs/plugins/custom_config.conf

qingling128 reviewed (commented):
Overall LGTM. Left some minor documentation / testing related comments. Most of them are not blocking this PR (I tried to make it clear by calling them out).

Commit:
check source header for apache 2.0 license (#4)

* check source header for apache 2.0 license

* use script to check for invalid headers

* on pr and commit to master

* resolve shellcheck concerns

* fix name

* install go and install addlicense with go install

* fix syntax

* add Google license headers

* Fix linux config checksums, they are different because of the license addition

Commit:
re-enable 1.0.4 ops agent tests on linux (#5)

Commit:
Script descriptions (#6)

* Add context to inspec and puppet scripts. Remove project and zone variables from inspec / puppet scripts as they are no longer used.

* Add context to setup.sh

* Add context to ssh.sh

* Add context to terraform wrapper scripts

* project is still required for the action

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
use package_state string instead of installed bool (#7)

Commit:
rename cloud_ops to google_cloud_ops (#8)

* rename cloud_ops to google_cloud_ops

* update module path to reflect new module name

* use latest checksum for config

Commit:
support 2.x.x (#9)

* try 2.0.0 in ci

* update test cases to support GA 2.0.0 release

* Remove support for pre ga

* update to ga service name

* Remove support for pre ga

* fix windows version typo

* test against new service names

* use new checksum for default config

* use 2.0.0 config

* use correct checksums

* switch to 2.0.1, resolve issue where systemd service is 'not running'

* use 2.0.1 for example version

* use 2.0.1 for example version

* fix config path

* fix config path

* fix checksum due to new path

* fix 2.0.1 service name

* temp disable some runs to save time

* detect service name based on ops agent version

* try 1.0.10 with 2.0.1 upgrade

* Add 1.0.10

* try split builtin

* remove 1.x.x support. test 2.0.1 on windows

* test windows 2.0.1

* enable all

Commit:
Update README.md

Co-authored-by: Ling Huang <qingling128@users.noreply.github.com>

Commit:
pr feedback (#10)

* pr feedback

* run ci on changes to metadata.json

Commit:
clarify packer images

Commit:
Logging additional config (#11)

* use custom config, not plugin

* test logging

* fix checksum for new test config

* re-enable all

qingling128 reviewed (approved):
LGTM

Commit:
move create image step to end

qingling128 commented:
Just merged this PR.

Looks like some integration tests failed. Might need some investigation.
![image](https://user-images.githubusercontent.com/5287526/125705847-8d458f63-1f72-435f-95fd-32f4e6bb7fe1.png)

BTW, I just noticed the last commit of the PR did not trigger integration tests: 
![image](https://user-images.githubusercontent.com/5287526/125705858-41d1d7f3-c49b-4ab8-adc0-38064cdcf311.png)

I guess some of the commits that addressed code review feedback introduced a breakage somewhere, while the file is not captured by: https://github.com/GoogleCloudPlatform/puppet-google-logging/blob/ff1be4759a86a4a46c8f2c5b7bf961fce017f0ec/.github/workflows/linux.yml#L5

On the side, we are also looking for Puppet experts inside Google to see if there are any Puppet conventions we need to follow within Google or in general. If there is any feedback there, we can send follow up PRs to address them.

jsirianni commented:
That is expected. The repo does not have secrets created. We should talk
about next steps for implementing CI against a GCP project.

On Wed, Jul 14, 2021, 19:36 Ling Huang ***@***.***> wrote:

> Just merged this PR.
>
> Looks like some integration tests failed. Might need some investigation.
> [image: image]
> <https://user-images.githubusercontent.com/5287526/125705847-8d458f63-1f72-435f-95fd-32f4e6bb7fe1.png>
>
> BTW, I just noticed the last commit of the PR did not trigger integration
> tests:
> [image: image]
> <https://user-images.githubusercontent.com/5287526/125705858-41d1d7f3-c49b-4ab8-adc0-38064cdcf311.png>
>
> I guess some of the commits that addressed code review feedback introduced
> a breakage somewhere, while the file is not captured by:
> https://github.com/GoogleCloudPlatform/puppet-google-logging/blob/ff1be4759a86a4a46c8f2c5b7bf961fce017f0ec/.github/workflows/linux.yml#L5
>
> On the side, we are also looking for Puppet experts inside Google to see
> if there are any Puppet conventions we need to follow within Google or in
> general. If there is any feedback there, we can send follow up PRs to
> address them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/puppet-google-logging/pull/8#issuecomment-880278651>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFPZ57AJOJURZCEYP4DWTW3TXYNPXANCNFSM47GPSMRQ>
> .
>
